### PR TITLE
refactor: improve readable helper boundaries

### PR DESCRIPTION
**Pull request by ben1009**
## Summary

Refactor several dense core paths into smaller helpers and clean up the readability-driven clippy fallout.

## Changes
- split long helper-heavy paths in `lsm_storage.rs`, `compact.rs`, `wal.rs`, `manifest.rs`, `table.rs`, `mem_table.rs`, `lsm_iterator.rs`, and `vlog/mod.rs`
- keep the normal `cargo make test` path focused on lib, bin, and test targets
- add a dedicated nextest `all-targets` profile that runs bench binaries serially with longer timeouts
- keep `cargo make test-all-targets` available for the full all-targets workflow

## Verification
- [x] `cargo make check`
- [x] `cargo nextest run --workspace --all-features --all-targets -E 'not kind(bench)'`
- [x] `cargo test --workspace --all-features --benches --no-run`
- [x] benchmark binaries start and list cases successfully (`deleterange_benchmarks`, `vlog_benchmarks`, `wal_bench`)

## Notes
- `--all-targets` under nextest was timing out because Criterion bench binaries were being run like normal tests; the new profile isolates those bench targets into a serial group with a long timeout.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved reliability and correctness across compaction, manifest recovery, write batching, MVCC/versioned reads, and vLog GC cleanup.
  * Fixed edge cases involving tombstones, exact-version lookups, range tombstones, and WAL recovery/validation to make reads and recovery more consistent.
* **Chores**
  * Updated test execution to better separate standard runs from full-target benchmark runs.
  * Added a dedicated benchmark test profile to run benchmark binaries serially with safer timeouts and no retries.

### CODE REVIEW

File: kv-engine/src/lsm_storage.rs [hunk 1]
```diff
@@ -2590,6 +2628,120 @@ impl LsmStorageInner {
...
+            return Ok(());
+        }
+
+        let seek_user_key = crate::key::encoded_user_key_prefix(seek)
+            .expect("encoded lower bound must have a valid user key prefix");
+        while iter.is_valid()
+            && crate::key::encoded_user_key_prefix(iter.key().raw_ref()) == Some(seek_user_key)
+        {
+            iter.next()?;
+        }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Fixed. `advance_past_lower_excluded` now handles the non-MVCC case explicitly: if the lower bound is not timestamp-encoded, it skips only exact key matches instead of calling `encoded_user_key_prefix(...).expect(...)`.

File: kv-engine/src/wal.rs [hunk 2]
```diff
@@ -1421,6 +1436,78 @@ impl Wal {
...
+        }
+
+        let mut state = self.completion_state.mutex.lock();
+        if let Err(e) = result {
+            state.last_error = Some(Arc::new(anyhow::anyhow!("{e}")));
+        } else {
+            state.last_error = None;
+            self.completion_state
+                .durable_ticket
+                .store(max_ticket + 1, Ordering::Release);
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/wal.rs]: Fixed. `publish_submit_result` now uses `durable_ticket.fetch_max(max_ticket + 1, Ordering::Release)` so the durable watermark only moves forward under concurrent leaders.

File: .config/nextest.toml [hunk 3]
```diff
@@ -12,3 +12,18 @@ slow-timeout = { period = "10s", terminate-after = 4, grace-period = "3s" }
...
+# assigned to a serial test group with a much longer timeout.
+[profile.all-targets]
+retries = { backoff = "exponential", count = 3, delay = "1s", jitter = true }
+test-threads = "num-cpus"
+
+[[profile.all-targets.overrides]]
+filter = 'kind(bench)'
+test-group = 'bench-serial'
+slow-timeout = { period = "10m", terminate-after = 1, grace-period = "10s" }
+retries = 0
```
[the remark this answers is not in the document]
Comment (ben1009) [reply .config/nextest.toml:26]: Not changed. This profile is intentionally specialized for the `test-all-targets` workflow, which includes Criterion bench binaries that do not behave like normal nextest targets. The standard repo policy remains in `profile.default`, and `cargo make test` still uses that fast default path. The bench-only override keeps regular tests on the normal settings while assigning `kind(bench)` to a serial group with a long timeout so the all-targets workflow remains usable.

**Comment by ben1009** [comment]
Addressed the Gemini and CodeRabbit findings on the latest head `5e5bcf5` and reran `cargo make check`.

@coderabbitai review
@gemini-code-assist please re-review the latest changes.

**Comment by ben1009** [comment]
@gemini-code-assist Addressed both re-review findings on `86c9c19`:

1. **wal.rs submit_and_commit**: Moved the `durable_ticket.load(Acquire) > ticket` fast-path check to the top of the function, before loading `next_ticket` or entering the CAS loop. Returns immediately when the ticket is already durable.

2. **lsm_storage.rs build_scan_range_tombstone_iterator**: Removed the redundant prefix bloom filter check — range tombstones are not indexed by prefix blooms. Replaced with a simple `if !sst.has_range_tombstones() { continue; }` that also avoids the double SST lookup. Dropped the now-unused `prefix_hint` parameter.

Please re-review.

File: kv-engine/src/wal.rs [hunk 4]
```diff
@@ -1318,12 +1389,7 @@ impl Wal {
...
-            let mut file = self.buffered_file.lock();
-            file.flush().context("failed to flush WAL")?;
-            file.get_ref()
-                .sync_all()
-                .context("failed to sync WAL to disk")?;
-            return Ok(());
+            return self.flush_legacy_wal();
         }
 
         let next_ticket = self.next_ticket.load(Ordering::Acquire);
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/wal.rs:1404]: Fixed. Moved the durable_ticket.load(Acquire) > ticket fast-path check to the top of submit_and_commit, immediately after the non-MVCC early-return. This avoids loading next_ticket, running the validation check, and entering the CAS loop when the ticket is already durable.

File: kv-engine/src/lsm_storage.rs [hunk 5]
```diff
@@ -2590,6 +2636,125 @@ impl LsmStorageInner {
...
+                && !sst.may_contain_prefix(prefix)
+            {
+                continue;
+            }
+            if let Some(sst) = state.sstables.get(id)
+                && let Some(frags) = sst.range_tombstone_fragments()
+                && !frags.is_empty()
+            {
+                lists.push(frags);
+            }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs]: Fixed. Removed the redundant prefix bloom filter check — range tombstones are not indexed by prefix blooms, so may_contain_prefix was irrelevant. Replaced with a simple if !sst.has_range_tombstones() guard that also avoids the double SST lookup. Dropped the now-unused prefix_hint parameter.

**Comment by ben1009** [comment]
@gemini-code-assist All 4 round-3 findings addressed on `8352cc4`:

1. **replay_range_only_ssts (Critical)**: Removed ro_ids_in_state blocks from Leveled and ForceFullCompaction arms — Leveled only compacts overlapping range-only SSTs, and ForceFull already fully replaces via std::mem::take.

2. **ForceFullCompaction redundant block**: Removed alongside #1.

3. **lookup_sst_raw_non_mvcc try_hint**: Added the O(1) hinted-SST optimization matching the MVCC variant. Avoids unnecessary partition_point binary search when the hinted SST covers the lookup key.

4. **get_versioned_raw_with_key_and_hash allocation**: Replaced per-call Vec::new() with thread_local RefCell<Vec<u8>> buffer reuse.

Please re-review.
